### PR TITLE
Feature/jsonapi (part 1)

### LINKED ISSUE #27: Support JSON API serialization PART 1
**Data**
- [x] Include `type` and `id`
- [x] Incluse attributes
- [x] Include relationships data without its attributes
- [x] Include array of relationships

### PR DESCRIPTION
**Pull request by MP0w**
Closes #27 

Should support all the basic stuff for JSON API 
Thanks @Landsi @stanislavstavrev and @joanromano for the help during DONI
Was quite easy to solve the empty array issue:
`Element.self is JSONAPISerializable.Type` // we just forgot `.Type`

### CODE REVIEW

File: Source/JSONAPISerialization.swift [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+     - parameter includeAttributes:    Defines if it should include the `attributes` field
+     
+     - returns: Return an object representing the `data` field conforming to JSON API
+     */
+    func data(includeRelationships includeRelationships: Bool, includeAttributes: Bool) -> AnyObject?
+}
+
+/**
+ *  A JSON API entity, conforming to this protocol will change the behavior of serialization, `CustomSerializable` behavior will be overriden by the JSON API behavior.
+ *  Relationships of an entity should also conform to this protocol, relationships are automatically recognized using the staic type of the property. For example an Array of `JSONAPIEntity` would be recognized as a relationship, also when empty, as soon as is static type at compile time is inferred correctly.
```
Comment (joanromano) [inline Source/JSONAPISerialization.swift]: `using the staTic type`

Comment (joanromano) [inline Source/JSONAPISerialization.swift]: I would also add a line break here, otherwise sentence becomes to large to read

File: Source/JSONAPISerialization.swift [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ 
+    ```swift
+        struct User: JSONAPIEntity {
+            let friends: [Friend] // correct if friend is JSONAPIEntity
+            let enemies: [Any] // incorrect, Any is not JSONAPIEntity so this property is an attribute instead of a relationship
+        }
+    ```
+ */
+public protocol JSONAPIEntity: CustomSerializable, JSONAPISerializable {
```
Comment (joanromano) [inline Source/JSONAPISerialization.swift]: Overall the explanations regarding JSONAPI are good, but I would also add some references to the official JSONAPI when possible (e.g for the `relationships` and `attributes` parameters, or how are the relationships included within data)

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            it("should not serialzie attributes of relationships") {
+                let object = json(user)
+                let dogData = object["relationships"]["dog"]["data"].dictionary
+                expect(dogData).toNot(beNil())
+                expect(dogData?["attributes"]).to(beNil())
+            }
+            
+            it("should not serialzie nil relationships") {
+                let object = json(dog)
+                let cat = object["relationships"].dictionaryValue
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: I don't get this: dog has a non-nil relationship in this case `let dog = Dog(id: "22", name: "Joan", chasingCat: cats[0])`

Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: And in this case probably the serialization is wrong then, because cat should be part of relationships with only `id` and `type`. I think `JSONAPISerializable` mirroring is not working well with non-nil optionals

Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: Serialization of `Dog(id: "22", name: "Joan", chasingCat: Cat(id: "33", name: "Stancho"))` adds cat as an attribute instead of a relationship because the mirroring check fails. 

Should we also handle optionals there?

Comment (MP0w) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: yes the test is completely wrong, and Optional handling for JSONAPIEntity is missing, adding now

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            
+            it("should fail to serialize CustomSerializable entities") {
+                // TODO: discuss because this might be unexpected
+                let object = json(CustomPost(id: "123", title: "Test"))
+                expect(object["id"].string).toNot(beNil())
+                expect(object["foo"].string).to(beNil())
+            }
+        }
+        
+        describe("JSON API Entity relationship serialization") {
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift:146]: Only inside this `describe`: change all `serialzie` with `serialize`

File: Source/JSONAPISerialization.swift [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        
+        data["id"] = id
+        data["type"] = type
+        
+        let mirror = Mirror(reflecting: self)
+        
+        var attributes = [String: AnyObject]()
+        var relationships = [String: AnyObject]()
+        
+        for child in mirror.children {
```
Comment (joanromano) [inline Source/JSONAPISerialization.swift]: We can avoid some mirroring when `includeRelationships` or `includeAttributes` are `false`

Comment (MP0w) [inline Source/JSONAPISerialization.swift]: true I forgot to add this at the DONI

File: Source/JSONAPISerialization.swift [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+            if let label = child.label {
+                if let value = child.value as? JSONAPISerializable, let data = value.data(includeRelationships: false, includeAttributes: false) {
+                    if includeRelationships {
+                        relationships[label] =  ["data": data]
+                    }
```
Comment (joanromano) [inline Source/JSONAPISerialization.swift]: Why don'y we avoid doing `value.data` if we already know that `includeRelationships` is `false`? Also, shouldn't we completely skip the mirroring iteration for non-root objects (`includeRelationships` and `includeAttributes` = `false`)?

Comment (MP0w) [inline Source/JSONAPISerialization.swift]: we can't avoid `value.data` because if it returns nil it means that value is an attribute (just added doc for this), we can avoid it only if `includeAttributes` is also nil but then we would never reach this point (when both boolean are nil).
For the second question I don't get it

Comment (joanromano) [inline Source/JSONAPISerialization.swift]: Ok, got your point on my first concern.

My second concern is: afaik, for non-root objects we only return `id` and `type`. Does `includeRelationships and includeAttributes = false` means that we are a non-root object? Can we skip all this `for child in mirror.children` then?

Comment (joanromano) [inline Source/JSONAPISerialization.swift]: Oh sorry didn't see the first `guard`

Nevermind then

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+                let object = json(lonelyMax)
+                let attributes = object["attributes"].dictionaryValue
+                expect(attributes.count).to(equal(1)) // only name should be here, no id, no cats
+                expect(attributes["name"]).to(equal("Max"))
+                expect(attributes["cats"]).to(beNil())
+                expect(attributes["id"]).to(beNil())
+            }
+            
+            it("should fail to serialize CustomSerializable entities") {
+                // TODO: discuss because this might be unexpected
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: I'm confused, why `customSerialize()` from `CustomPost` is not called? Shouldn't it override the default implementation in the extension? 

Anyway, if it's never called, shouldn't we add an `XCTFail()` there?

Comment (MP0w) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: I'm also not sure why exactly.... yes we can add XCTFail()

Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: Still don't get why it's not called

Comment (MP0w) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: I'm an idiot... it was called before but then the behavior changed... the reason is that I didn't update it to return `AnyObject?`

Comment (MP0w) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: fixed in 9f10b22

Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: Ok then now this is the expected behavior isn't it? If you override `customSerialize` you will get back your own implementation.

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+                let lonelyMax = User(id: "11", name: "Max", dog: dog, cats: [])
+                let object = json(lonelyMax)
+                let cats = object["relationships"]["cats"].dictionary!
+                expect(cats.count).to(equal(1))
+                let dataArray = cats["data"]
+                expect(dataArray).toNot(beNil())
+                expect(dataArray?.count).to(equal(0))
+            }
+            
+            context("Optional relationships") {
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift:197]: Didn't know about `context`, Quick has so many stuff 💃

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+                it("should not be included when nil") {
+                    let object = json(Dog(id: "123", name: "A", cat: nil))
+                    let relationships = object["relationships"].dictionaryObject
+                    expect(relationships).to(beNil())
+                }
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: This test is mostly the same as `should not serialize nil relationships` I think, but I'm fine with leaving both

File: KakapoExample/KakapoTests/JSONAPITests.swift (hunk 7 above)
Comment (MP0w) [inline KakapoExample/KakapoTests/JSONAPITests.swift]: yes and the test was like this before I changed it

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+                let data = object["relationships"]["policy"].dictionaryObject!["data"] as? [String: AnyObject]
+                expect(data).toNot(beNil())
+                expect(data?.count).to(equal(0))
+            }
+            
+            it("should exclude PropertyPolicy as releantionships when the associated type is JSONAPIEntity but .None") {
+                let object = json(Policy<User>(id: "12", policy: .None))
+                let relationships = object["relationships"].dictionary
+                expect(relationships).to(beNil())
+            }
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift:266]: Only in this describe: change `releantionships` with `relationships`

File: KakapoExample/KakapoTests/JSONAPITests.swift [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+            
+            it("should serialize an arrays of non-JSONAPIEntities as an attribute") {
+                let object = json(Post(id: "11", relatedPostIds: [1, 2, 3]))
+                let relatedPostIds = object["attributes"]["relatedPostIds"].arrayValue
+                expect(relatedPostIds.count).to(equal(3))
+                expect(relatedPostIds[0]).to(equal(1))
+                expect(relatedPostIds[1]).to(equal(2))
+                expect(relatedPostIds[2]).to(equal(3))
+            }
```
Comment (joanromano) [inline KakapoExample/KakapoTests/JSONAPITests.swift:112]: Probably add another test with empty arrays of non-JSONAPIEntities

File: Source/JSONAPISerialization.swift [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        case let .Some(value):
+            if let value = value as? JSONAPISerializable {
+                return value.data(includeRelationships: includeRelationships, includeAttributes: includeAttributes)
+            }
+            
+        case .Null:
+            return [String: AnyObject]() // included as relationship but empty
+        case .None:
+            return nil
+        }
```
Comment (joanromano) [inline Source/JSONAPISerialization.swift]: `Null` returns empty dict and `None` nil, a bit confusing for me after reading `PropertyPolicy` docs

File: Source/JSONAPISerializer.swift [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+/**
+ *  A JSON API entity, conforming to this protocol will change the behavior of serialization, `CustomSerializable` entities should not adopt this protocol because the default implementation would be overridden by their custom serialization.
+ *  Relationships of an entity should also conform to this protocol.
+ *  Properties recognized as relationships:
+    - conforming to `JSONAPIEntity`
+    - Array of `JSONAPIEntity`
+    - Optional `JSONAPIEntity`
+    - PropertyPolicy `JSONAPIEntity`
+ 
+ * In general to be recognized as potential relationship or relationship wrapper (like Array or Oprionals) an object must conform to `JSONAPISerializable` and implement its methods.
```
Comment (joanromano) [inline Source/JSONAPISerializer.swift]: `Optionals`

**Comment by joanromano** [comment]
👍
